Fix: enable flags needed to run tsgen js output based on settings

Dabolus opened the pull request:
This draft PR aims to solve the issues reported on https://github.com/emscripten-core/emscripten/issues/20514 and https://github.com/emscripten-core/emscripten/issues/20542.

It's still a WIP and requires a discussion to understand what would be the best way to solve that issues.

### Alternative solution

The changes in this PR aim to make the TS embind generation as close as possible to the JS/WASM generation by leaving all the flags enabled. However, an easier solution would probably be to simply disable those flags when generating the typings, as the TypeScript output is exactly the same.

brendandahl commented on src/embind/embind_ts.js:
```diff
@@ -210,6 +210,8 @@ var LibraryEmbind = {
         ['bool', 'boolean'],
         ['float', 'number'],
         ['double', 'number'],
+        ['long', 'number'],
```
When `MEMORY64` is enabled these will be bigint.

brendandahl reviewed (changes requested):
Please add a test case to https://github.com/emscripten-core/emscripten/blob/main/test/other/embind_tsgen.cpp and update the reference file https://github.com/emscripten-core/emscripten/blob/main/test/other/embind_tsgen.d.ts under normal mode. We'll also want to add separate test for memory64 mode similar to what you did before.

sbc100 commented on emcc.py:
```diff
@@ -3307,7 +3307,14 @@ def phase_embind_emit_tsd(options, in_wasm, wasm_target, memfile, js_syms):
+  # Build the flags needed by Node.js to properly run the output file.
+  node_args = []
+  if settings.MEMORY64:
+    node_args += ['--experimental-wasm-memory64']
```
We have `shared.node_memory64_flags()` for this.

We don't have the same thing for wasm-eh since it looks like we only allow that to run on node 17 and above: https://github.com/emscripten-core/emscripten/blob/aa250479f654fb8dffb0ff69132da1779128da99/test/common.py#L695-L701.   This suggests to me that the wasm-eh support in node 16 doesn't work, even with the `--experimental-wasm-eh`.   @aheejin do you know if this is true?

Dabolus replied:
@sbc100 updated the code to use the shared helper. As for the exception handling, given the fact that it shouldn't have any impact in the generated typings (correct?), would it make sense to just disable it as done with other options?

sbc100 replied:
Assuming exceptions can be disabled at link time like this then yes that sounds good.

Dabolus replied:
I did this change and the tests seem to pass ✅

aheejin replied:
> We have `shared.node_memory64_flags()` for this.
> 
> We don't have the same thing for wasm-eh since it looks like we only allow that to run on node 17 and above:
> 
> https://github.com/emscripten-core/emscripten/blob/aa250479f654fb8dffb0ff69132da1779128da99/test/common.py#L695-L701
> 
> . This suggests to me that the wasm-eh support in node 16 doesn't work, even with the `--experimental-wasm-eh`. @aheejin do you know if this is true?

Sorry for the delayed reply. #18654 says v16 didn't support the newly added `{traceStack}` option used for stack traces.

Dabolus replied:
Ok then, do you think it would be safer to return an error when trying to generate typings with a Node.js version that doesn't properly support wasm exceptions (i.e. < 17)?

Dabolus commented:
I added the mapping so that the types are generated as bigints when MEMORY64 is enabled. I also added the related test and updated the documentation accordingly. I think this PR should be ready now 👍🏻
